File: core/baker-test/src/test/java/com/ing/baker/test/javadsl/RecipeAssertTest.java
package com.ing.baker.test.javadsl;

import com.ing.baker.runtime.javadsl.EventInstance;
import com.ing.baker.runtime.common.BakerException.NoSuchProcessException;
import com.ing.baker.test.EventsFlow;
import com.ing.baker.test.RecipeAssert;
import com.ing.baker.test.recipe.WebshopBaker;
import com.ing.baker.test.recipe.WebshopRecipe;
import org.junit.Test;
import scala.collection.JavaConverters;
import scala.collection.mutable.Buffer;

import java.util.Arrays;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;

import static org.junit.Assert.assertEquals;

public class RecipeAssertTest {

    private RecipeAssert createRecipeAssert(int delayInSeconds) {
        String recipeInstanceId = UUID.randomUUID().toString();
        WebshopBaker.javaBaker().bake(WebshopBaker.recipeId(), recipeInstanceId);
        Buffer<String> items = JavaConverters.asScalaBuffer(Arrays.asList("item-1", "item-2", "item-3"));

        if (delayInSeconds > 0) {
            new Timer().schedule(
                    new TimerTask() {
                        @Override
                        public void run() {
                            WebshopBaker.javaBaker().fireEvent(recipeInstanceId,
                                    EventInstance.from(new WebshopRecipe.OrderPlaced("order-1", items.toList())));
                        }
                    },
                    delayInSeconds * 1000L
            );
        } else {
            WebshopBaker.javaBaker().fireEvent(recipeInstanceId,
                    EventInstance.from(new WebshopRecipe.OrderPlaced("order-1", items.toList())));
        }

        return RecipeAssert.of(WebshopBaker.javaBaker(), recipeInstanceId);
    }

    private RecipeAssert createRecipeAssert() {
        return createRecipeAssert(0);
    }

    @Test
    public void testHappy() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventsFlow(WebshopRecipe.happyFlow());
    }

    @Test
    public void testEventNamesHaveHappened() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventNamesHappened(WebshopRecipe.happyFlow().events().last());
    }

    @Test
    public void testEventsHaveHappened() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventsHappened(WebshopRecipe.ItemsReserved.class);
    }

    @Test(expected = AssertionError.class)
    public void testEventNamesHaveNotHappened() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventNamesNotHappened("NonExistentEvent", "NonExistentEvent2");
    }

    @Test(expected = AssertionError.class)
    public void testEventsHaveNotHappened() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventsNotHappened(Object.class);
    }

    @Test
    public void testHappyAfterDelay() {
        createRecipeAssert(2)
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventsFlow(WebshopRecipe.happyFlow());
    }

    @Test(expected = AssertionError.class)
    public void testFailWithIncorrectEvents() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertEventsFlow(WebshopRecipe.happyFlow().remove(WebshopRecipe.OrderPlaced.class));
    }

    @Test
    public void testAssertIngredientIsEqual() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("orderId").isEqual("order-1");
    }

    @Test(expected = AssertionError.class)
    public void testAssertIngredientIsEqualFail() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("orderId").isEqual("order-2");
    }

    @Test
    public void testAssertIngredientIsAbsent() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("not-existing").isAbsent();
    }

    @Test(expected = AssertionError.class)
    public void testAssertIngredientIsNull() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("orderId").isNull();
    }

    @Test
    public void testAssertIngredientCustom() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("orderId").is(value -> assertEquals("order-1", value.as(String.class)));
    }

    @Test(expected = AssertionError.class)
    public void testAssertIngredientCustomFail() {
        createRecipeAssert()
                .waitFor(WebshopRecipe.happyFlow())
                .assertIngredient("orderId").is(value -> assertEquals("order-2", value.as(String.class)));
    }

    @Test(expected = NoSuchProcessException.class)
    public void testNoSuchProcessFailure() {
        EventsFlow flow = EventsFlow.of("Event");
        RecipeAssert.of(WebshopBaker.javaBaker(), UUID.randomUUID().toString())
                .waitFor(flow);
    }
}

File: core/baker-test/src/test/scala/com/ing/baker/test/scaladsl/RecipeAssertTest.scala
package com.ing.baker.test.scaladsl

import cats.effect.IO
import cats.effect.unsafe.IORuntime
import com.ing.baker.runtime.scaladsl.EventInstance
import com.ing.baker.test.RecipeAssert
import com.ing.baker.test.recipe.WebshopBaker._
import com.ing.baker.test.recipe.WebshopRecipe
import com.ing.baker.test.recipe.WebshopRecipe.{ItemsReserved, OrderPlaced}
import com.typesafe.scalalogging.StrictLogging
import org.scalatest.Assertions
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

import java.util.UUID
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.util.{Failure, Try}

class RecipeAssertTest extends AnyFlatSpec with Matchers with StrictLogging {

  implicit val runtime: IORuntime = IORuntime.global

  private def createBakerAssert(async: Boolean = false): IO[RecipeAssert] = {
    val recipeInstanceId = UUID.randomUUID().toString
    val sensoryEvent = EventInstance.unsafeFrom(OrderPlaced("order-1", "item-1" :: Nil))

    // Helper to convert Baker's Future to IO
    def futureToIO[A](f: => Future[A]): IO[A] = IO.fromFuture(IO(f))

    val fireEventAction = futureToIO {
      baker.fireSensoryEventAndAwaitReceived(recipeInstanceId, sensoryEvent)
    }

    for {
      _ <- futureToIO(baker.bake(recipeId, recipeInstanceId))
      _ <- if (async) IO.sleep(500.millis) *> fireEventAction else fireEventAction
    } yield {
      RecipeAssert(baker, recipeInstanceId)
    }
  }

  private def assertFail[T](assertion: => T): Unit = Try(assertion) match {
    case Failure(t: Throwable) => logger.info(s"Assertion failed as expected: ${t.getMessage}")
    case default => fail(s"assertion error is expected but got $default")
  }

  "RecipeAssert object" should "be created" in {
    RecipeAssert.apply(baker, "someProcessId")
  }

  "assertEventsFlow" should "work with happy flow" in {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow)
  }

  "assertEventsFlow" should "work with some delay" in {
    val recipeAssert = createBakerAssert(true).unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow)
  }

  "assertEventsFlow" should "fail on incorrect events" in assertFail {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow -- classOf[ItemsReserved])
  }

  "assertIngredient" should "work for isEqual" in {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").isEqual("order-1")
  }

  "assertIngredient" should "fail for isEqual" in assertFail {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").isEqual("order-2")
  }

  "assertIngredient" should "work for isAbsent" in {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("non-existent").isAbsent
  }

  "assertIngredient" should "fail for isNull" in assertFail {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").isNull
  }

  "assertIngredient" should "work for customAssert" in {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").is(value => Assertions.assert(value.as[String] == "order-1"))
  }

  "assertIngredient" should "fail for customAssert" in assertFail {
    val recipeAssert = createBakerAssert().unsafeRunSync()
    recipeAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").is(value => Assertions.assert(value.as[String] == "order-2"))
  }
}